File: mcp-client-bridge/src/routing/routing-table.ts
/**
 * Fetches, caches, and refreshes the tool routing table from the orchestrator.
 * Supports ETag caching and periodic refresh.
 */

import { BridgeConfig } from '../bridge-config.js';

export interface ToolRoute {
  location: 'local' | 'remote';
  server?: string;
  fallback?: 'local' | 'remote';
  priority?: number;
}

export interface RoutingTableData {
  version: string;
  updatedAt: string;
  defaultLocation: 'local' | 'remote';
  tools: Record<string, ToolRoute>;
}

const EMPTY_TABLE: RoutingTableData = {
  version: '0.0.0',
  updatedAt: '',
  defaultLocation: 'remote',
  tools: {},
};

export class RoutingTable {
  private cached: RoutingTableData = EMPTY_TABLE;
  private etag: string | null = null;
  private refreshTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly refreshIntervalMs: number;
  private readonly baseUrl: string;
  private readonly token: string | null;

  constructor(config: BridgeConfig, refreshIntervalMs = 60_000) {
    this.baseUrl = config.orchestratorUrl;
    this.token = config.token;
    this.refreshIntervalMs = refreshIntervalMs;
  }

  /** Get the cached routing table. */
  getCached(): RoutingTableData { return this.cached; }

  /** Resolve a tool name to its route. */
  resolve(toolName: string): ToolRoute | null {
    return this.cached.tools[toolName] ?? null;
  }

  /** Get the default location when tool is not in table. */
  get defaultLocation(): 'local' | 'remote' {
    return this.cached.defaultLocation;
  }

  /** Set routing table from initialize response _meta. */
  setFromMeta(meta: Record<string, unknown>): void {
    const rt = meta.routingTable as RoutingTableData | undefined;
    if (rt && rt.tools && typeof rt.tools === 'object') {
      this.cached = rt;
      console.error(`[routing-table] Loaded from _meta (${Object.keys(rt.tools).length} routes)`);
    }
  }

  /** Fetch routing table from orchestrator with ETag caching. */
  async fetch(): Promise<boolean> {
    try {
      const headers: Record<string, string> = { 'Accept': 'application/json' };
      if (this.token) headers['Authorization'] = `Bearer ${this.token}`;
      if (this.etag) headers['If-None-Match'] = this.etag;

      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 10_000);

      const res = await fetch(`${this.baseUrl}/api/routing-table`, {
        method: 'GET',
        headers,
        signal: controller.signal,
      });
      clearTimeout(timeout);

      if (res.status === 304) {
        return true; // Not modified, cache is valid
      }

      if (!res.ok) {
        console.error(`[routing-table] Fetch failed: HTTP ${res.status}`);
        return false;
      }

      const data = await res.json() as RoutingTableData;
      if (!data.tools || typeof data.tools !== 'object') {
        console.error('[routing-table] Malformed response: missing tools');
        return false;
      }

      this.cached = data;
      this.etag = res.headers.get('etag');
      console.error(`[routing-table] Updated (${Object.keys(data.tools).length} routes, v${data.version})`);
      return true;
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      console.error(`[routing-table] Fetch error: ${msg} (keeping cache)`);
      return false;
    }
  }

  /** Start periodic refresh. */
  startRefresh(): void {
    if (this.refreshIntervalMs <= 0) return;
    this.scheduleRefresh();
  }

  /** Stop periodic refresh. */
  stopRefresh(): void {
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
      this.refreshTimer = null;
    }
  }

  /** Trigger an immediate refresh (e.g., after local server restart). */
  async refresh(): Promise<void> {
    await this.fetch();
  }

  private scheduleRefresh(): void {
    this.refreshTimer = setTimeout(async () => {
      await this.fetch();
      this.scheduleRefresh();
    }, this.refreshIntervalMs);
  }
}

File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/routing/RoutingTable.kt
package com.orchestrator.mcp.bridge.routing

import com.orchestrator.mcp.bridge.BridgeConfig
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import org.slf4j.LoggerFactory

/**
 * Routing definition for a single tool.
 */
data class ToolRoute(
    val location: String, // "local" or "remote"
    val server: String? = null,
    val fallback: String? = null,
    val priority: Int = 0,
)

/**
 * Full routing table data from orchestrator.
 */
data class RoutingTableData(
    val version: String = "0.0.0",
    val updatedAt: String = "",
    val defaultLocation: String = "remote",
    val tools: Map<String, ToolRoute> = emptyMap(),
)

/**
 * Fetches, caches, and refreshes the tool routing table from the orchestrator.
 * Supports ETag caching and periodic refresh.
 */
class RoutingTable(
    private val config: BridgeConfig,
    private val refreshIntervalMs: Long = 60_000,
) {
    private val logger = LoggerFactory.getLogger(RoutingTable::class.java)
    private val json = Json { ignoreUnknownKeys = true }
    private var cached = RoutingTableData()
    private var etag: String? = null
    private var refreshJob: Job? = null

    private val httpClient = HttpClient(CIO) {
        engine { requestTimeout = 10_000 }
    }

    /** Get the cached routing table. */
    fun getCached(): RoutingTableData = cached

    /** Resolve a tool name to its route. */
    fun resolve(toolName: String): ToolRoute? = cached.tools[toolName]

    /** Get the default location when tool is not in table. */
    val defaultLocation: String get() = cached.defaultLocation

    /** Set routing table from initialize response _meta. */
    fun setFromMeta(meta: JsonObject) {
        val rt = meta["routingTable"]?.jsonObject ?: return
        cached = parseTable(rt)
        logger.info("Loaded from _meta ({} routes)", cached.tools.size)
    }

    /** Fetch routing table from orchestrator with ETag caching. */
    suspend fun fetch(): Boolean {
        return try {
            val response = httpClient.get("${config.orchestratorUrl}/api/routing-table") {
                header(HttpHeaders.Accept, "application/json")
                config.token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
                etag?.let { header(HttpHeaders.IfNoneMatch, it) }
            }

            if (response.status == HttpStatusCode.NotModified) return true
            if (!response.status.isSuccess()) {
                logger.warn("Fetch failed: HTTP {}", response.status.value)
                return false
            }

            val body = json.parseToJsonElement(response.bodyAsText()).jsonObject
            if (body["tools"] == null) {
                logger.warn("Malformed response: missing tools")
                return false
            }

            cached = parseTable(body)
            etag = response.headers[HttpHeaders.ETag]
            logger.info("Updated ({} routes, v{})", cached.tools.size, cached.version)
            true
        } catch (e: Exception) {
            logger.warn("Fetch error: {} (keeping cache)", e.message)
            false
        }
    }

    /** Start periodic refresh. */
    fun startRefresh(scope: CoroutineScope) {
        if (refreshIntervalMs <= 0) return
        refreshJob = scope.launch { refreshLoop() }
    }

    /** Stop periodic refresh. */
    fun stopRefresh() {
        refreshJob?.cancel()
        refreshJob = null
    }

    /** Trigger an immediate refresh. */
    suspend fun refresh() { fetch() }

    private suspend fun refreshLoop() {
        while (currentCoroutineContext().isActive) {
            delay(refreshIntervalMs)
            fetch()
        }
    }

    private fun parseTable(obj: JsonObject): RoutingTableData {
        val tools = mutableMapOf<String, ToolRoute>()
        obj["tools"]?.jsonObject?.forEach { (name, element) ->
            val route = element.jsonObject
            tools[name] = ToolRoute(
                location = route["location"]?.jsonPrimitive?.contentOrNull ?: "remote",
                server = route["server"]?.jsonPrimitive?.contentOrNull,
                fallback = route["fallback"]?.jsonPrimitive?.contentOrNull,
                priority = route["priority"]?.jsonPrimitive?.intOrNull ?: 0,
            )
        }
        return RoutingTableData(
            version = obj["version"]?.jsonPrimitive?.contentOrNull ?: "0.0.0",
            updatedAt = obj["updatedAt"]?.jsonPrimitive?.contentOrNull ?: "",
            defaultLocation = obj["defaultLocation"]?.jsonPrimitive?.contentOrNull ?: "remote",
            tools = tools,
        )
    }
}
